src/Core/CustomRedirects/CustomRedirectCollection.cs

```diff
@@ -14,12 +17,45 @@ public class CustomRedirectCollection: CollectionBase
         /// <summary>
         /// Hashtable for quick lookup of old urls
         /// </summary>
-        private readonly Hashtable _quickLookupTable = new Hashtable(StringComparer.InvariantCultureIgnoreCase);
+        private readonly Dictionary<string, CustomRedirect> _quickLookupTable = new Dictionary<string, CustomRedirect>();
+
+        /*[CanBeNull]*/
+        private string GetLookupKey(/*[NotNull]*/ CustomRedirect customRedirect)
+        {
+            if (customRedirect == null) throw new ArgumentNullException(nameof(customRedirect));
+            return GetLookupKey(customRedirect.OldUrl);
+        }
+
+        /*[CanBeNull]*/
+        private string GetLookupKey(/*[CanBeNull]*/ string url)
+        {
+            return UrlStandardizer.Standardize(url);
+        }
+
+        private void AddLookup(/*[NotNull]*/ CustomRedirect customRedirect)
+        {
+            if (customRedirect == null) throw new ArgumentNullException(nameof(customRedirect));
+
+            _quickLookupTable[GetLookupKey(customRedirect)] = customRedirect;
+        }
+
+        private void RemoveLookup(/*[NotNull]*/ CustomRedirect customRedirect)
+        {
+            _quickLookupTable.Remove(GetLookupKey(customRedirect));
+        }
+
+        /*[CanBeNull]*/
+        private CustomRedirect TryLookup(/*[NotNull]*/ string url)
+        {
+            CustomRedirect foundRedirect = null;
```

Comment:
can you use here new C# feature to inline `out` variable? Can't remember whether we have enabled latest C# tho..